### ISSUE TITLE
lldb does not support "<vector>" and "<union>" when describing registers

### ISSUE THREAD

**Issue by DavidSpickett** [llvm/llvm-project#87471]
This was reported to me by someone trying to debug SVE and SME code that switched streaming modes, using qemu-user. When in streaming mode they could not read the Z registers but they could in non-streaming mode. Qemu was returning valid data each time but lldb refused to display it in streaming mode.

The cause is a difference in how lldb-server and qemu describe the type of these registers. lldb-server describes them as:
```
<reg name="z31" bitsize="256" regnum="193" encoding="vector" format="vector-uint8" group="Scalable Vector Extension Registers" ehframe_regnum="127" dwarf_regnum="127" />
```
And this gets converted to `eEncodingVector`. We can see that lldb-server sets that in the source:
```
// Defines a Z vector register with 16-byte default size
#define DEFINE_ZREG(reg)                                                       \
  {                                                                            \
    #reg, nullptr, 16, 0, lldb::eEncodingVector, lldb::eFormatVectorOfUInt8,   \
        SVE_REG_KIND(reg), nullptr, nullptr, nullptr,                          \
  }
```
This format lets us handle any amount of elements in the vector. So if you increase the vector length, no problem.

qemu on the other hand uses `<vector>` and `<union>` (see https://sourceware.org/gdb/current/onlinedocs/gdb.html/Target-Description-Format.html):
```
<vector id="svevqu" type="uint128" count="16"/>
<...>
</union><union id="svev">
  <field name="q" type="svevnq"/>
  <field name="d" type="svevnd"/>
  <field name="s" type="svevns"/>
  <field name="h" type="svevnh"/>
  <field name="b" type="svevnb"/>
</union>
<reg name="z0" bitsize="2048" regnum="34" type="svev"/>
```
lldb doesn't recognise the type `svev` and so defaults to unsigned int. That goes through https://github.com/llvm/llvm-project/blob/e05c1b46d0d3739cc48ad912dbe6e9affce05927/lldb/source/Utility/RegisterValue.cpp#L189 which only supports up to 128 bit numbers. So you can read the Z registers if the vector length is less than 128 bits (vg = 2) but as soon as it goes beyond that, lldb rejects the data.

The example I was given used streaming mode and set qemu's SVE/SME default vector lengths like:
```
-cpu max,sve-default-vector-length=16,sme-default-vector-length=64
```
But you can also reproduce by writing vg from within lldb `register write vg 4`.

We could add some arbitrary sized integer support (`Scalar` already stores the value using `APSInt` and `APFloat`). This would be the quick fix.

Long term we should support `<vector>` and `<union>` as they do have some advantages. `<union>` causes the debugger to display the vector in multiple ways, which can be useful for SVE where you don't know what type is in the register at a given point.

**Comment by DavidSpickett** [comment]
For `<union>` we can model this as a C union type and it would work for unions of `<flags>` types. For vectors if they are sized we could make C types as arrays but if they are scalable then that's more tricky. One option is to convert them into lldb's preferred formatting options.

Also now I see that these union types are sized, so I wonder what happens when the vector length changes. It's possible qemu changes the representation, which we won't pick up because we I don't think we re-parse all the XML.

**Comment by DavidSpickett** [comment]
Currently the C struct types we make have a big endian layout, and we compensate for that by swapping the over of the fields *in the value* if the target's endian is not big.

This presents a problem for union of flags. If two union members are flags with different layouts, we can't supply one value that's the correct order for both.

This might extend to union of vector as well, but I'm not about to melt my mind trying to reason about that yet.

So although the GDB spec does not prohibit nesting these types arbitrarily, we may have to add some restrictions just to make it implementable.

**Comment by DavidSpickett** [comment]
One option here might be to define the structs in an order that matches the endian of the value that we have, then print the struct members in reverse order. Then the one value contained in the union can be correct for all of the flags members.

**Comment by DavidSpickett** [comment]
So far I got Neon working:
```
(lldb) register read v0
      v0 = {0x00 0x01 0x02 0x03 0x04 0x05 0x06 0x07 0x08 0x09 0x0a 0x0b 0x0c 0x0d 0x0e 0x0f}
         = {
             d = {
               f = (7.9499288951273625E-275, 3.6919162048650923E-236)
               u = (506097522914230528, 1084818905618843912)
               s = (506097522914230528, 1084818905618843912)
             }
             s = {
               f = (3.82047143E-37, 1.00825135E-34, 2.65846276E-32, 7.00365327E-30)
               u = (50462976, 117835012, 185207048, 252579084)
               s = (50462976, 117835012, 185207048, 252579084)
             }
             h = {
               u = (256, 770, 1284, 1798, 2312, 2826, 3340, 3854)
               s = (256, 770, 1284, 1798, 2312, 2826, 3340, 3854)
             }
             b = {
               u = (0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f)
               s = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
             }
             q = {
               u = (20011376718272490338853433276725592320)
               s = (20011376718272490338853433276725592320)
             }
           }
```
(and found https://github.com/llvm/llvm-project/issues/107864 along the way)

SVE is still not working because the initial read fails due to the > 128 bit fallback issue.

**Comment by satyajanga** [comment]
@DavidSpickett Thank you for filing and keeping the issue upto date. We are also running into similar issue at Meta. 

Do you have a plan to upstream the commits in your fork?

**Comment by DavidSpickett** [comment]
Yes I have implementations of union and vector ready to be upstreamed once I have finished with https://github.com/llvm/llvm-project/pull/189590. My goal is to have all this in upstream LLDB eventually.

**Comment by satyajanga** [comment]
let me know if there is a way for us to help speed up this. @barsolo2000  and I are available. 
@clayborg  can help support this too. !

**Comment by barsolo2000** [comment]
@DavidSpickett if you want to take a look here #195887 (to be honest I did it before finding out you were working on it). 
It should be built on top of #189590 (If I understand correctly.. you are fixing the LLDB has a "hack" that builds the Clang bitfield struct with fields in the wrong positions, then reverses the register value before display to compensate (ReverseFieldOrder)).

**Comment by DavidSpickett** [comment]
Yes I'm fixing that hack, because:
* I know for sure that it does not work for a `union` of many `flags` (though it's probably not an issue for `union` of many `vector`).
* It will make it easy(er) to remove the current requirement that register size be 4 or 8. Should be able to handle any size.
* Putting the types in the right AST can help expression support (it's not a requirement for now).

I took a brief look at your draft and I have some issues with the code organisation but on the other hand you've gone further in some respects like resolving the basic type names and supporting union of concrete types (I only implemented union of other XML defined types).

My draft version is https://github.com/llvm/llvm-project/pull/196032.

My main issue with your version is adding a pointer for the union type next to the flags type. In my version I've added a base type for the register types which solves that.

So assuming your use case isn't effected by the field reverse hack, I could upstream the refactoring parts and you can get on with your work based on that. While I deal with the field hack / AST situation. Sound good to you?

**Comment by barsolo2000** [comment]
> Yes I'm fixing that hack, because:
> 
> * I know for sure that it does not work for a `union` of many `flags` (though it's probably not an issue for `union` of many `vector`).
> * It will make it easy(er) to remove the current requirement that register size be 4 or 8. Should be able to handle any size.
> * Putting the types in the right AST can help expression support (it's not a requirement for now).
> 
> I took a brief look at your draft and I have some issues with the code organisation but on the other hand you've gone further in some respects like resolving the basic type names and supporting union of concrete types (I only implemented union of other XML defined types).
> 
> My draft version is [#196032](https://github.com/llvm/llvm-project/pull/196032).
> 
> My main issue with your version is adding a pointer for the union type next to the flags type. In my version I've added a base type for the register types which solves that.
> 
> So assuming your use case isn't effected by the field reverse hack, I could upstream the refactoring parts and you can get on with your work based on that. While I deal with the field hack / AST situation. Sound good to you?

lets do that!

**Comment by DavidSpickett** [comment]
Great, I will start putting up the refactoring parts then, since I don't know when I'll be able to figure out the other bits.

**Comment by clayborg** [comment]
It would be great if we can create clang AST types from the <union ...> and <vector ...> and <flags ...> and then we provide the data for the register in the register context for the top level register type, and our existing code for ValueObject can do all of the work after that. This also means that the registers would work in expressions. 

Which leads to a feature we might be able to do is to have a type defined with source code instead of some <union ...> <vector ...> and <flags ...> custom representation. It would be great if we would specify any source code along with a language type that would produce a type we can feed to the expression parser and extract. 

```
<source lang="c++" typename="<typename>">
union <typename> {
 float f;
 double d;
};
</source>
```

Where `<typename>` is the name of the type to use for any registers.

**Comment by clayborg** [comment]
But it should be really easy to convert any <union ...> and <vector ...> and <flags ...> to type system CompilerType objects.

**Comment by DavidSpickett** [comment]
Yes I was able to do it. See RegisterTypeBuilderClang.cpp - https://github.com/llvm/llvm-project/pull/196032/changes#diff-63c45b7dab4534531cf720867ab684660e321010dc82926568aa146a3f4f28b0

**Comment by barsolo2000** [comment]
@DavidSpickett  Hi David, its been a while. I was wondering where do we stand with the changes you wanted to make, so I can rebase with https://github.com/llvm/llvm-project/pull/195887 and create the Union.

**Comment by DavidSpickett** [comment]
I replied on the PR, paraphrasing that here: an alternative approach to the refactoring was proposed. I have time to upstream my refactoring changes but I do not have time to re-engineer them to the alternative approach.

**Comment by DavidSpickett** [comment]
The stack ending at https://github.com/llvm/llvm-project/pull/213897 has all the refactoring I want to do.

From there, you can go ahead with the rest. https://github.com/llvm/llvm-project/pull/196032/commits has some code for vector and union if you want to reuse any of it.

I don't think I'm going to fix my union of flags problem any time soon, but it isn't used in the field yet anyway. So if you need to hard code handling for certain type combinations, or just reject them, that's fine.

For the purposes of this issue the most important part is a union of vector where the lanes are basic types, and that's possible without fixing the flags problem.

**Comment by DavidSpickett** [comment]
With https://github.com/llvm/llvm-project/pull/223485 we can parse the unions:
```
$ bin/lldb -o "settings set platform.plugin.qemu-user.architecture aarch64" -o "platform select qemu-user" -o "target create /tmp/test.o" -o "b main" -o "run" -o "register info z0"
<...>
(lldb) register info z0
       Name: z0
       Size: 64 bytes (512 bits)
Invalidates: v0, s0, d0
    In sets: general (index 0)

  Union members:
    q (svevnq, 256 bytes)
    d (svevnd, 256 bytes)
    s (svevns, 256 bytes)
    h (svevnh, 256 bytes)
    b (svevnb, 256 bytes)
```
That only shows one level of the union but it looks correct and the logs imply we're parsing the inner parts too.
```
lldb             ProcessGDBRemote::ParseRegisters Size of register type svep (256 bytes) for register p0 does not match the register size (32 bytes). Ignoring this type.
```
This is probably due to qemu sending us the max SVE size rather than the current.

**Comment by DavidSpickett** [comment]
With https://github.com/llvm/llvm-project/pull/224147, we can display the types!

I have to set the vector length to max to match what qemu claims, but that's a different issue. The program sets each Z register to a pattern of [lane number, register number].

```c
#include <sys/prctl.h>
#include <linux/prctl.h>
#include <stdio.h>
#include <stdlib.h>

#define SVE_REGS(M) \
  M(0)  M(1)  M(2)  M(3)  M(4)  M(5)  M(6)  M(7)  \
  M(8)  M(9)  M(10) M(11) M(12) M(13) M(14) M(15) \
  M(16) M(17) M(18) M(19) M(20) M(21) M(22) M(23) \
  M(24) M(25) M(26) M(27) M(28) M(29) M(30) M(31)

#define INIT_Z(N) \
  "index z" #N ".h, #0, #1\n" \
  "add z" #N ".h, z" #N ".h, #" #N ", lsl #8\n"

int main(void) {
  if (prctl(PR_SVE_SET_VL, 256) == -1) {
    perror("PR_SVE_SET_VL");
    return 1;
  }

  asm volatile(
      SVE_REGS(INIT_Z)
      :
      :
      : "memory");

  // Break here.
  asm volatile("nop");

  return 0;
}
```

```
$ ninja && ./bin/lldb -o "settings set platform.plugin.qemu-user.architecture aarch64" -o "platform select qemu-user" -o "target create /tmp/test.o" -o "b main" -o "run"
<...>
(lldb) n
<...>
(lldb) n
<...>
(lldb) register read z0
  z0 = 0x007f007e007d007c007b007a00790078007700760075007<...>
     = (q = (u = (36346553384763581871632909800112128, <...>)), d = (f = (4.1720559249406128E-309, <...>), u = (844433520132096, <...goes on for a long time...>)))
(lldb) register info z0
       Name: z0
       Size: 256 bytes (2048 bits)
Invalidates: v0, s0, d0
    In sets: general (index 0)

  Union members:
    q (svevnq, 256 bytes)
    d (svevnd, 256 bytes)
    s (svevns, 256 bytes)
    h (svevnh, 256 bytes)
    b (svevnb, 256 bytes)
(lldb) register read z0.b.u
  z0.b.u = (0x00, 0x00, 0x01, 0x00, 0x02, 0x00, 0x03, 0x00, 0x04, 0x00, 0x05, 0x00, 0x06, 0x00, 0x07, 0x00, 0x08, 0x00, 0x09, 0x00, 0x0a, 0x00, 0x0b, 0x00, 0x0c, 0x00, 0x0d, 0x00, 0x0e, <...>)
(lldb) register read --format uint8_t[] z0
  z0 = {0x00 0x00 0x01 0x00 0x02 0x00 0x03 0x00 0x04 0x00 0x05 0x00 0x06 0x00 0x07 0x00 0x08 0x00 0x09 0x00 0x0a 0x00 0x0b 0x00 0x0c 0x00 0x0d 0x00 0x0e <...>}
```

Seems correct too. Massive amount of data which comes with the territory of vectors, but being able to narrow it own with `.` is super cool.

Seriously impressive work @barsolo2000 !
